### PR TITLE
Added option to remove Modal header

### PR DESCRIPTION
**Pull request by HardeepAsrani**
Added an option which allows you to remove header from the modal, that addresses: #13207

### CODE REVIEW

**Review by jorgefilipecosta — COMMENTED** [review]
Hi @HardeepAsrani thank you for your PR. In my tests, it worked as expected 👍
I left minor suggestions for possible code improvements.
I agree with the idea of having an option to disable header on the modal, but I'm not sure about the potential UX implications it would be awesome if we have more thoughts on this cc: @jasmussen.

File: packages/components/src/modal/index.js [hunk 1]
```diff
@@ -144,14 +145,16 @@ class Modal extends Component {
...
-							headingId={ headingId }
-							icon={ icon }
-							isDismissable={ isDismissable }
-							onClose={ onRequestClose }
-							title={ title }
-						/>
+						{ ! noHeader &&
+							<ModalHeader
+								closeLabel={ closeButtonLabel }
+								headingId={ headingId }
```
[the remark this answers is not in the document]
Comment (HardeepAsrani) [reply packages/components/src/modal/index.js]: For that, just a small if condition to only calculate the ID when noHeader is false? Or should we do something else?

**Comment by HardeepAsrani** [comment]
For the potential UX implications, in the most ideal scenario, the developer is supposed to keep the modal header. The option is when someone wants to write their own modal header to have a better UX according to its needs.

I can't think of a major UX implication but you know better so would love to learn what issues it can cause. Also, I'm little worried about how this header-less behavior is, in terms of a11y? :)

**Comment by jasmussen** [comment]
Thank you for your contribution! 

There are two key concerns here, and the most important one being accessibility. In short, it has to be possible for a blind person to open the modal, have a screenreader announce the title of the modal, and use the tab key both to navigate the modal and close it again. If the modal is opened without header, presumably this doesn't work (I took the PR for a spin but wasn't able to get it to work, that's my fault not yours). While it would probably be possible for a plugin author to create a new header, as you suggest in the ticket, which was equally accessible, we cannot be sure that a plugin author will actually do this, and we might end up with reduced accessibility in some plugins as a result, even if unintentionally.

One of the greatest benefits of Gutenberg is that it is component based. This allows us to design great components with accessibility built in, that are reused by WordPress itself, as well as plugins, extending accessibility beyond just the core offering. Given that accessibility is such a key value to WordPress, this is an important aspect to consider. 

In that vein, I would rather we discuss what prompted this desire in the first place, and explore ways to extend the API to address those problems. For example, I would very much like to see an `isCompact` property that would allow you to open a tiny and minimalist modal. I could imagine other properties that might help you accomplish what you want, without removing and rewriting the header.

So in that vein: what header are you trying to build?

**Comment by HardeepAsrani** [comment]
This is the type of header that we're building. It's accessible but you're right that not everyone is going to build with this in mind.

It can be different in each use case, but it will give designers the flexibility to make things, but I agree that a11y should be **the priority** when doing such changes.

Tried a few different designs with tabs below the header but it didn't look good. You can close the modal without a close button too with ESC key if that prop is set to true.

To me, the close thing isn't an issue as there's already remove the close button, and the modal title is already optional, not required. So is it making the modal any less accessible than a developer not removing close button and removing the title?

Thank you for testing, and would love to see discussion how we can approach this with accessibility in mind.

screen shot 2019-01-08 at 11 16 29 pm

**Comment by jasmussen** [comment]
Thank you so much for sharing that mockup, it is incredibly helpful in order to figure out how to accommodate this "tabbed modal" use case.

As mentioned, one of the benefits to having a standard API for modals, is that WordPress itself can handle accessibility, and in this case also responsiveness. By using the standard modal you can be sure that it works well on mobile, and indeed when further improvements are added, those will trickle to the API as well.

In that vein, I think it's extra important that we await a good accessibility review before proceeding further with this PR (though thank you so much for doing the work already!). 

However one path forward would be to change #13207 (or close that ticket and create a new?) to be about adding support for tabbed modals. It seems a fair enough use case, and by building that into the API we'd ensure that it was built in an accessible and mobile friendly way. 

In the mean time, I imagine through clever CSS, you could probably achieve the use case suggested in that mockup. By creating two navigation links and two sections as the content of your modal, you could add CSS to pull up those two navigation links and make them tabs. The modal could be slimmed down using CSS only — so it seems you can accomplish your goal without actually removing the header markup from the source. Though I would recommend ensuring accessibility and mobile friendliness as you do.

Was this helpful?

**Comment by HardeepAsrani** [comment]
> However one path forward would be to change #13207 (or close that ticket and create a new?) to be about adding support for tabbed modals.

I don't think a Tabbed Modal should be an option, as it's too niche and not many people will be using it, so it shouldn't be in the core. The idea here is to make Modal more flexible rather than adding different types of modal.

> In the mean time, I imagine through clever CSS, you could probably achieve the use case suggested in that mockup.

I've already achieved it, so the issue is really not for our use-case, but for future if people want to be creative and make an awesome looking modal which will put my modal design to shame. 😄

> Though I would recommend ensuring accessibility and mobile friendliness as you do.

Yes, that's my point. We can think over an option to remove header, if we can do it while making sure a11y rules are followed, else I'll be happy to close the pull request.

Do you think we should wait for someone from a11y team and see what they say, and close the pull request if it's not possible with a11y in mind? 😄

**Comment by jasmussen** [comment]
Thank you for the additional context, and I'm glad you managed to solve it. I also deeply appreciate that you were willing to donate this PR to the project, in case others wanted to do something similar.

> Do you think we should wait for someone from a11y team and see what they say, and close the pull request if it's not possible with a11y in mind? 😄

Yes, I think we need to make sure this can work in an accessible way. I'm almost sure it can't, but I'd like a sanity check. @tofumatt I know you're a busy man, can you take a glance? Thank you 🧡

**Comment by jasmussen** [comment]
Thank you for your contributions, let's close this for now and look at other ways to enable this approach!

**Pull request closed without merging**